README.md
[<img src="https://upload.wikimedia.org/wikipedia/commons/thumb/a/ae/Flag_of_the_United_Kingdom.svg/100px-Flag_of_the_United_Kingdom.svg.png" height="30">](README.md)
[<img src="https://upload.wikimedia.org/wikipedia/commons/thumb/a/a4/Flag_of_the_United_States.svg/100px-Flag_of_the_United_States.svg.png" height="30">](README.md)
[<img src="https://upload.wikimedia.org/wikipedia/commons/thumb/b/ba/Flag_of_Germany.svg/100px-Flag_of_Germany.svg.png" height="30">](README.de_DE.md)

# Inofficial modification of the RF firmware for RF1000 and RF2000 devices
Based on Repetier-Firmware - the fast and user friendly firmware.

## Disclaimer

THIS VERSION IS AN UNOFFICIAL MODIFICATION OF mhiers UNOFFICIAL MODIFICATION OF THE ORIGINAL FIRMWARE. It is neither endorsed nor
supported by the developers of the RF1000 firmware.  
USE AT YOUR OWN RISK.

For the official version, have a look at the upstream repository:
https://github.com/RF1000/Repetier-Firmware (see branch development)

## HowTo Install

- Download the Firmware `Branch: community_development` and unzip all the files.  
- Install Arduino.cc 1.6.5 or 1.8.3 and later if it is not installed on your computer already.  
- Edit and save Configuration.h @Line46 and 47 according to your printers model. You have to remove the two **//** in front of the printers name you wish to activate:  
`#define MOTHERBOARD                         DEVICE_TYPE_RF1000` or  
`#define MOTHERBOARD                         DEVICE_TYPE_RF2000`
- Optional: Edit and save the Configuration.h and RF1000.h or RF2000.h to your needs in case you use special specialized hardware.
- Connect your Printer with a USB-cable and switch it on.
- Go to and doubleclick /Repetier/Repetier.ino
- Within the cyan colored Arduino-IDE you have to choose some preconfiguration settings for your RFx000 Board at the menu **Tools->**:  
Board: `Arduino Mega 2560 or Mega ADK`,  
Processor: `ATMega 2560 (Mega 2560)`,  
Port: Your printers Port like `COM3`, if successfully connected.
- **Check** and **Upload** the Firmware to your Printer.
## If you upgrade to this Version from 1.37r or earlier please do a fresh M303 PID-Autotune on all heaters!  
- RF2000/RF1000 extruder 1: `M303 P0 X0 S230 R10 J1`
- RF2000 extruder 2: `M303 P1 X0 S230 R10 J1`
- RF2000 heated bed: `M303 P2 X0 S70 R15 J3`
- RF1000 heated bed: `M303 P1 X0 S70 R15 J3`  
And start with EEPROM-values `PID drive min = 5` and `PID drive max = 100`  
Or set #define PID_CONTROL_DRIVE_MIN_LIMIT_FACTOR to 10.0f in configuration.h to use the old version PID-control.

## Version RF.01.37mod 

See:  
http://www.rf1000.de/viewtopic.php?f=74&t=1674 (Nibbels/Wessix SenseOffset-Thread)  
http://www.rf1000.de/viewtopic.php?f=7&t=1504#p14882 (mhier Mod)  
http://www.rf1000.de/viewtopic.php?f=7&t=1504&start=60#p16397 (added feature)  
http://www.rf1000.de/viewtopic.php?f=7&t=1610 (rf1000s developement version thread to 1.35)  
http://www.rf1000.de/viewtopic.php?f=7&t=1665 (rf1000s developement version thread to 1.37)  
https://github.com/Nibbels/Repetier-Firmware/commits/heat_bed_z_offset_scan-%26-heat_bed_decrease (all the commits)

## Main Features of this Mod:
* mhier Z-Offset-Scan
* Nibbels/Wessix SensiblePressure Adjustment

## List of Features and additional G-Codes for RF2000 and RF1000

_by mhier_:  
M3900               - Scan the heatbeds offset and correct the active zMatrix within the RAM of the Printer. (X-Y-Pos, Learning-Factor, linear distance weight.)  
RF2000: Z_OVERRIDE_MAX = 1.0mm, instead of 0.5mm  
RF1000: Z_OVERRIDE_MAX = 1.0mm, instead of 0.5mm  

_by StarTony_:  
Fan-Speed-Patch, see http://www.rf1000.de/viewtopic.php?f=7&t=1638  

_by Wessix_:  
M355 Sx - Turn on and off, or switch Port X19 to controll lights etc. (24V-MOSFET @Connector X19)  
M355    - Switch Port X19 to controll lights etc. (24V-MOSFET @Connector X19)  
Idea to implement the SensiblePressure-Function, see M3909  
Idea to send the active digits with the printTemperature-Status.

_by Nibbels_:  
M3900 Xn Ym Px Sy   - Scan the heatbeds offset and correct the active zMatrix within the RAM of the Printer. (X-Y-Pos, Learning-Factor, linear distance weight.)  
M3901 Xn Ym Px Sy   - to preconfigure M3900 (X-Y-Pos, Learning-Factor, linear distance weight.)  

M3902 R1            - to fix a single HeatBeds Hole within the HBS Matrix.  
M3902 Zn.n          - to add an Offset to the Matrix. n.n = {-0.2 .. 0.2} [mm]  
M3902 Z0            - to shift your active zOffset to the zMatrix. The M3006 zOffset will be zero afterwards. The zMatrix is altered within Ram and might be saved to EEPROM[n] with M3902 S[n] afterwards.  
M3902 Sn            - to save the active Matrix to position n = {1..9}  
M3902 Z0 S1         - to shift the zOffset to your zMatrix and save the Matrix at position 1. This is an example to show that the options of M3902 can be combined.  
M3939 Fn St1 Pt2 Ex Iy Rm   - to messure a curve of your Filaments velocity over digits -> viscosity.  

M3920 Sb - to go into or switch back from SilentMode (This will lower your steppers current to an alternative profile)  
Removed all compiler warnings and errors within the original Firmware.  
Included some Fixes the original developers of the branch "repetier/Repetier-Firmware" committed to their firmware.  
Upgraded the Firmware to the latest RF.01.37 (2017-01-20)  
Activated the X35 Temperature-Sensor for RF2000.  
Listed all spare pins for RF1000 or RF2000.  
Fixed the RF2000's Status Message length to 20chars.  
Fix for M3117
Third Z-Scale Mode: G-Code @ Configuration->Z-Configuration->Z-Scale.

_by Nibbels/Wessix_:  
M3909 Pn Sm         - See "Nibbels/Wessix SenseOffset"  

## mhier Z-Offset-Scan
* M3900 Xn Ym Sy Px             - Run a Z-Offset Scan at the specified Location.

Preconfiguration Options for M3900:  
* M3901 Xn Ym Sy Px 

[X]/[Y] specifies the Location of the Scan within the original HBS-Locations.  
X = {1..9}, {0 = random}
Y = {2..12}, {0 = random}
With two extruders X=1 is allowed otherwise forbidden. Normally X=2 is the lowest possible scan position. This limit is caused by border values within every zMatrix. You cannot set it wrong, the upper and lower limits are limited and autoadjusted by the firmware if set wrong.  

[S] specifies the learning rate of M3900  
S= {100}  
The measured offset will be added/substracted by 100% of its Value from/to the original matrix. The original matrix is the matrix stored in your EEPROM. Thus the matrix will be reloaded from the EEPROM to the RAM prior to the following z-offset scan.
S= {0..99}  
the Matrix in RAM will be adjusted by 0% to 99% of the messured offset. This is a great feature for multiple incremental corrections, in case you would not want to start allover or you cannot trust your first values 100% (you want to sum up corrections from different scanning-locations).

[P] ZOS learning linear distance weight  
P = {0..100}  
This is a configuration option to commit a scans offset to its surrounding area only.  
If you scan the middle of the heated bed and then all the corners with a high distance weight, this feature will behave like autobed leveling of your zMatrix.
Example: Put P to a high value and messure all corners. Then the matrix is somehow "bended" and not constantly updated. (Might work very well on heat-beds which tend to lift edges within different temperature ranges or to correct tilting).

You can activate the mhier Z-Offset-Scan within ->Configuration->Z-Configuration->Z-Offset Scan within your printers menu structure. This will always use the preconfigured or changed (see M3901) settings within your printers RAM. These M3900/M3901-Settings are not stored within the printers EEPROM.  

## Nibbels/Wessix SenseOffset  
* M3909 Pn Sm  
P = max. digits = {1...14999} [digits]  
S = max. SenseOffset = {1...300} std.: 180 [um]  

Feature called "Nibbels/Wessix SenseOffset". This is a digit limiter which decreases the printers digits by lowering the heatbed. It compensates the thermal expansion of some second order printer parts.  
 The printer will automaticly release Pressure inbetween the Nozzle and the HeatBed, whenever the digits rise atop [max.digits] by adjusting the Z-Offset. 
 The adjustment-offset is fixed to positive values (bed does never go closer to the nozzle than without M3909). The adjustment will not violate the "max. SenseOffset" restriction in order to avoid unwanted offsets (not caused by too close nozzle).  
 When choosing "max. SenseOffset" you should already know how much digits your printer normally has (Same Material, same Temperatures, same Speed, same Nozzle) and then add some plus 20% or plus 1000digits of force-tolerance.  
 Senseoffset can have two states. Active controling and idle. It is idle, whenever Z is not within negative and ~0.3mm height. You can tune this height limit by tweaking M3007 (http://www.rf1000.de/wiki/index.php/GCodes#M3007_-_Minimale_H.C3.B6he_f.C3.BCr_Z-Kompensation_festlegen). Whenever there is active control you see a circumflex "^" within the modmenu, instead of the @-Symbol.
 ![ ](https://downfight.de/picproxy.php?url=https://downfight.de/grafiken/dimi/modmenudisplay_eng.png "Angebratenes ABS")  
 When the z-compensation gets deactivated the senseoffset-function is deactivated as well.  
 Use "M3909 P0" for manual shutdown of the feature, but normally this is not necessary.  
 You cannot activate the Feature if zCompensation is not active already.  
 
 While Printing you can adjust or activate this feature with the "Left"- and "Right"-Button, while being in the "Mod-Menu" which is page 2 on the printers display.  

## SensibleViscosity  
* M3939 Fn St1 Pt2   Ex Iy Rm  
S = Starting Temperature for the Test in [°C]  
P = Final Temperature for the Test in [°C]  
F = max. test digits = 10000 or {1...12000} [digits]  
E = max. test extrusion velocity = 5 or {0.05...maxStartFeedrate} [mm/s]  
optional:  
I = test velocity increment = 0.1 or {0.02...0.4} [mm/s]  
R = max. digit change needed to consider having a filled nozzle. = 800 [digits] 

Feature called "SensibleViscosity"  
 The printer will automaticly extrude into the thin air and messure the resulting force. 
 After every messurement the velocity of extrusion is increased.  
 You will get a CSV-Output within your Console-Window (Repetierserver/Repetierhost/Octoprint/...) which you can easily copy and paste to some text file. Excel can then read the Data and you can put it into some Chart with 2 Clicks.
 You might ask why you need this feature, but in case you know your Filament quite well you can compare these resulting chars to really see how your filament extrusion behaves under different circumstances or when your nozzle is or gets clogged.
 For longer heated/boiled (my) ABS: http://www.rf1000.de/viewtopic.php?f=62&t=1549&p=16617#p16617  
 ![ ](https://downfight.de/picproxy.php?url=http://image.prntscr.com/image/3508699873f7431489d1df9344110e7f.png "Boiled ABS")  
 For the general viscosity of (my) ABS: http://www.rf1000.de/viewtopic.php?f=23&t=1620&p=16613#p16605  
 ![ ](https://downfight.de/picproxy.php?url=http://image.prntscr.com/image/610de7720945474590668b20c05d7652.png "Temperature ABS")  
 That is a clogged nozzle with the same ABS:
 ![ ](https://downfight.de/picproxy.php?url=http://image.prntscr.com/image/c00ac11b3a384994b37ce8bc3cf03bd9.png "Clogged Nozzle ABS")  
 That is red PLA:
 ![ ](https://downfight.de/picproxy.php?url=http://image.prntscr.com/image/2a3253c930794afc81e4fa4d4b2a4261.png "Clogged Nozzle ABS")  

 Have fun finally seeing your Filaments behaviour over time, temp, nozzles ;)  
 
## SilentMode   
* M3920 Sb  
S = {0,1} Switch to SilentMode or back to normal stepper current and unhome your printer.  

This Feature might be handy if you want to predefine another set of stepper current, but you dont want to have this lower values all the time. 

If you make your current too low you might risk "lost steps". The Motor cannot stand the applied force and you might get shifted Layers (X,Y). If that is the case you should lower your acceleration or rise up the current again.

If you can lower your current, you will have a huge improvement on stepper temperature and noise production as well. The tone of your steppers might sound deeper and will not be such a pain as the original (milling like-) configuration.
You should not include M3920 when your z-Compensation is active. This MCode will shut down z-Compensation and unhome your Printer. You would have to re-home your axes and activate z-Compensation again. I put this code to the beginning of my startcodes.

## Dualhead Tip-Down Support (beta)
* M3919 [S]mikrometer - Testfunction for Dip-Down-Hotend @ T1: T1 can now be springloaded and the bed will be adjusted "down" whenever T1 is selected. This is alike Ultimaker 3 does it with the right hotend.   
Example: M3919 T1 Z-0.6 tells the Printer that the right hotend will reach 0.6mm more down than the left hotend. Test it :)

## Feature Digit-Z-Kompensation (beta)
If you apply force to the DMS sensors the bed is adjusted (if Z-Compensation is active). This corrects the error which bending of the sensors apply to your layers. The physical change of the beds level is approximately +0,01mm for +1000 applied force (digits).

## RF2000: Additional Temperature Sensor
This optional 3rd temperature T3 is automatically sent out with the other Temperatures and Digits.  
``` 15:16:14.637: T:204.67 /205 B:28.60 /20 B@:0 @:143 T0:28.60 /0 @0:0 T1:204.67 /205 @1:143 T3:28.23 F:322 ```
See RESERVE_ANALOG_SENSOR_TYPE in RF2000.h for configuration options.  
See RESERVE_ANALOG_INPUTS to find the involved code within the firmware.  
You will need to install and additional cable and a temperature sensor at X35 on your printers Board. Then you can wire your sensor to some place needed.  
This sensor might keep track of your boards temperature or the air-temperature within your enclosure.  

## Wessix`s help video:
[![ScreenShot](https://downfight.de/picproxy.php?url=http://image.prntscr.com/image/d7b7fade0c7343eeb67b680339478894.png)](http://youtu.be/iu9Nft7SXD8)

## !! 03.02.2017: Project is Work in Progress and untested changes are possible.
## !! 28.02.2017: This mod should be save to complile with Arduino 1.8.1. Former problems are gone.


changelog.txt
TODO: Pause-Back-Drive-Bug in original RF1.37
TODO: Possible Z-Lift Bug in original RF1.37

V RF.01.37r6.Mod (2017-08-01)
- Emergency-Stop for Z-Moves is now configurable within EEPROM and Printers Menu
- Text for RGB-Light is now "Auto" instead Color anymore. 
- Cleaned up EEPROM text for RGB-Light to get better Compatibility with Repetier-Server (Value is not hidden anymore)
- New Standard PID-Parameters for V2 because the old ones are quite bad in context to the improved new PID control.

V RF.01.37r5.Mod (2017-07-31)
- "PID drive max = 40"-Fix.
- Removed FEATURE_CONFIGURABLE_HOTEND because it was bogus on Dual and not really missed at RF1000 Single setup.
- Extruder Sensor Type is now configurable within EEPROM
- Cleanup, deleted unnecessary text and defines
- Added Thermistor-Types description as Comment to RFx000.h configurations

V RF.01.37r4.Mod (2017-07-29)
- Bugfix for Z-Matrix-Invalidation
- Extruder 2 Z-Offset got eeprom support
- Cleanup

V RF.01.37r3.Mod (2017-07-24)
- Fix M3902 Z0 to update Z-Offset in Flash too.
- Fix min-max-bug in Emergency-Pause Menu

V RF.01.37r2.TDMod (2017-07-24)
- Last SenseOffset Digits Limit is now stored in EEPROM. If you use a plain M3909 instead of M3909 Pxxxxx the last value is used from EEPROM automatically.
- The Emergency-Pause limit values are now editable and stored in EEPROM. EEPROM values are loaded at the printers start and can be updated via M3075.
- Emergency-Pause is now editable within Menu at Configuration->General (See at V 0.91.15 (2014-03-26) vor Details.)
- Updated small things according to repetier/Repetier @output EEPROM and some cleanup.
- Updated PID D-Gains temperature-diff to 1500ms timespan. This smoothes out the steering.

V RF.01.37r.TDMod (2017-07-22)
- Grouped my good old Auto-Matrix-Leveling to one Command: M3900 R1. This is available in Menu also. Configuration -> Z-Configuration -> "Auto-Matrix-Leveling"
- Z-Step-Size which can be set in Positon-Menu is now saved within flash automatically.
- Previous used maximal SenseOffset-Offsetlimit is now saved within flash automatically.
- Previous Z-Offset-Scan testpoints (X/Y) are now saved within flash automatically.
- Little Homing-Bug is gone (g_abortZScan might have been set to 1 if average digits read failed, which is problematic when using ZOffsetScan)

V RF.01.37q6.TDMod (2017-07-21)
- Testcase for possible endZCompensationStep-Bug in Z-Compensation.
- Testcase for possible PrintLine::direct.isZMove()-NoReturn-Bug in Z-Compensation.

V RF.01.37q5.TDMod (2017-07-17)
- Added J6=P-control, J5=PI-control, J4=PD-control or testing purposes to the PID autotune settings.

V RF.01.37q4.TDMod (2017-07-17)
- Fixing PID autotune for "some overshoot" by factor 2.0 like https://en.wikipedia.org/wiki/Ziegler%E2%80%93Nichols_method says.
- Added PID autotune ruleset called "Pessen Integral Rule" position J1 (See http://www.mstarlabs.com/control/znrule.html for tuning rulesets and description)

V RF.01.37q3.TDMod (2017-07-17)
- Upgrading PID Code support AutoTune with two additional presents: 3="No Overshoot" and 2="Some Overshoot". Use M303 with J[0..3]. Default is 0 and this means ClassicPID which was standard in all previous firmware versions.
- I switched the calculation of PID drive min to Minus. This should finally enable the real PID control - fix it!
- Speed-Mode for Menu-Klicking: Whenever you press some Button on the printer the Routine to answer is at 10ms clock instead of 100ms clock for 15 seconds. This enables us to navigate very fast now.

V RF.01.37q2.TDMod (2017-07-16)
- Upgrading PID Code to original repetier/Repetier :: added max-cycles
- Removed debug-watchdog-time-counter

V RF.01.37q.TDMod (2017-07-15)
- Implemented "Feature Zero Digits". When you home all Axes, the digits are messured and written to the digits offset. Then we see very little digits while printing. That should make them more comparable.
- Disabled FREE-RAM debugging (to not waste calculate power anymore on this)

V RF.01.37p6.TDMod (2017-07-13)
- Fixed Emergency-Stop to not make watchdog-reset
- Added Auto-Reset when the printer shows Emergency-Stop (Z-Block) and you press OK or PLAY.
- You cannot drive X- and Y-Axis out of bounds anymore using Menu "POSITION -> X/Y-Position -> ". This will now show an additional Warning if you aim out of bounds and the printer blocks it.
- Closing down ZOS-Debugging: ZOS now starts at 1mm Height. (Formerly 0.5mm, but 1mm spacing seems rather ok. 5mm was too high and slow. Spacing <= +12mm technically works god as of 01.37p5)

V RF.01.37p5.TDMod (2017-07-04)
- Fixed theoretical ZOS error with g_abortZScan, because when it is set by function of ZOS the function of HBS will abort without being active.

V RF.01.37p4.TDMod (2017-06-30)
- Improved repeating messurements within ZOS when the digits get disturbed somehow (Filament, Vibrations,...). ZOS is now much more robust against moving idle pressures.
- Test: Round Z-Slow-Steps to even Microsteps.

V RF.01.37p3.TDMod (2017-06-30)
- Tracing some ZOS-Bug with Filament
- Cleanup
- Fixed Bug in SchraubenJustierHelper: Da wurden bei der Abgrenzung mm mit Mikrometer verwechselt.

V RF.01.37p2.TDMod (2017-06-27)
- Fixed some details within ZOS-Scan
- Screw-Offset-Helper is now displaying its suggestions to the Printers display

V RF.01.37p.TDMod (2017-06-27)
- homing is now split into 3 axes
- homing Z only with Tipdown/Positive Matrix activated now might home Y first to avoid crash. This is experimental. This is not protecting the nozzle/bed very good.
- Small Z-Compensation Fix for TipDown-Support
- Tipdown for T1 is now adjustable within Menus Configuration -> Z-Configuration -> Extruder T1 TipDown -> +-
- activating Z-Compensation with only one axis homed is impossible now.
- whenever some axis is homed it is now shown with a * within menu.
- small cleanup

V RF.01.37o.TDMod (2017-06-20)
- Third Z-Scale: G-Code.
- Some small tweaks and comment updates to Tipdown-Support

Repetier/Eeprom.h
/*
    This file is part of the Repetier-Firmware for RF devices from Conrad Electronic SE.

    Repetier-Firmware is free software: you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation, either version 3 of the License, or
    (at your option) any later version.

    Repetier-Firmware is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with Repetier-Firmware.  If not, see <http://www.gnu.org/licenses/>.
*/


#ifndef EEPROM_H
#define EEPROM_H

// Id to distinguish version changes
#define EEPROM_PROTOCOL_VERSION         8

/** Where to start with our datablock in memory. Can be moved if you
have problems with other modules using the eeprom */


#define EPR_MAGIC_BYTE                  0
#define EPR_ACCELERATION_TYPE           1
#define EPR_XAXIS_STEPS_PER_MM          3
#define EPR_YAXIS_STEPS_PER_MM          7
#define EPR_ZAXIS_STEPS_PER_MM          11
#define EPR_X_MAX_FEEDRATE              15
#define EPR_Y_MAX_FEEDRATE              19
#define EPR_Z_MAX_FEEDRATE              23
#define EPR_X_HOMING_FEEDRATE_PRINT     27
#define EPR_Y_HOMING_FEEDRATE_PRINT     31
#define EPR_Z_HOMING_FEEDRATE_PRINT     35
#define EPR_MAX_JERK                    39
#define EPR_MAX_ZJERK                   47
#define EPR_X_MAX_ACCEL                 51
#define EPR_Y_MAX_ACCEL                 55
#define EPR_Z_MAX_ACCEL                 59
#define EPR_X_MAX_TRAVEL_ACCEL          63
#define EPR_Y_MAX_TRAVEL_ACCEL          67
#define EPR_Z_MAX_TRAVEL_ACCEL          71
#define EPR_BAUDRATE                    75
#define EPR_MAX_INACTIVE_TIME           79
#define EPR_STEPPER_INACTIVE_TIME       83
#define EPR_EXTRUDER_SPEED              95
#define EPR_INTEGRITY_BYTE              104   // Here the xored sum over eeprom is stored
#define EPR_VERSION                     105   // Version id for updates in EEPROM storage
#define EPR_BED_HEAT_MANAGER            106
#define EPR_BED_DRIVE_MAX               107
#define EPR_BED_PID_PGAIN               108
#define EPR_BED_PID_IGAIN               112
#define EPR_BED_PID_DGAIN               116
#define EPR_BED_PID_MAX                 120
#define EPR_BED_DRIVE_MIN               124
#define EPR_PRINTING_TIME               125  // Time in seconds printing
#define EPR_PRINTING_DISTANCE           129  // Filament length printed
#define EPR_X_HOME_OFFSET               133
#define EPR_Y_HOME_OFFSET               137
#define EPR_Z_HOME_OFFSET               141
#define EPR_X_LENGTH                    145
#define EPR_Y_LENGTH                    149
#define EPR_Z_LENGTH                    153
#define EPR_BACKLASH_X                  157
#define EPR_BACKLASH_Y                  161
#define EPR_BACKLASH_Z                  165

#if FEATURE_SERVICE_INTERVAL
#define EPR_PRINTING_TIME_SERVICE       169  // Time in seconds printing since last service intervall
#define EPR_PRINTING_DISTANCE_SERVICE   173  // Filament length printed since last service intervall
#define EPR_MILLING_TIME_SERVICE        177  // Time in seconds milling since last service intervall
#endif // FEATURE_SERVICE_INTERVAL

#define EPR_MILLING_TIME                181  // Time in seconds milling

#define EPR_RF_BEEPER_MODE              1000
#define EPR_RF_CASE_LIGHT_MODE          1001
#define EPR_RF_230V_OUTPUT_MODE         1002
#define EPR_RF_OPERATING_MODE           1003
#define EPR_X_HOMING_FEEDRATE_MILL      1004
#define EPR_Y_HOMING_FEEDRATE_MILL      1008
#define EPR_Z_HOMING_FEEDRATE_MILL      1012
#define EPR_RF_Z_ENDSTOP_TYPE           1016
#define EPR_RF_HOTEND_TYPE              1017 //depreached 30 07 2017 Nibbels : Die Funktion war nur für RF1000, nur Single, in der Falschen Klasse, wenn nötig, dann in Class Extruder einbauen. Hat beim RF2000 keiner vermisst: Ersatz ist EPR_EXTRUDER_SENSOR_TYPE
#define EPR_RF_MILLER_TYPE              1018
#define EPR_RF_RGB_LIGHT_MODE           1019
#define EPR_RF_FET1_MODE                1020
#define EPR_RF_FET2_MODE                1021
#define EPR_RF_FET3_MODE                1022
#define EPR_RF_RGB_PRINTING_R           1023
#define EPR_RF_RGB_PRINTING_G           1024
#define EPR_RF_RGB_PRINTING_B           1025
#define EPR_RF_RGB_HEATING_R            1026
#define EPR_RF_RGB_HEATING_G            1027
#define EPR_RF_RGB_HEATING_B            1028
#define EPR_RF_RGB_COOLING_R            1029
#define EPR_RF_RGB_COOLING_G            1030
#define EPR_RF_RGB_COOLING_B            1031
#define EPR_RF_RGB_IDLE_R               1032
#define EPR_RF_RGB_IDLE_G               1033
#define EPR_RF_RGB_IDLE_B               1034
#define EPR_RF_RGB_MANUAL_R             1035
#define EPR_RF_RGB_MANUAL_G             1036
#define EPR_RF_RGB_MANUAL_B             1037
#define EPR_RF_Z_OFFSET                 1038
#define EPR_RF_MOVE_MODE_X              1042
#define EPR_RF_MOVE_MODE_Y              1043
#define EPR_RF_MOVE_MODE_Z              1044
#define EPR_RF_Z_MODE                   1045 //das ist byte, 1046 wäre frei.

//vorsicht bei mehrbytigen Variablen, dann freilassen.

#define EPR_RF_MOD_Z_STEP_SIZE            1900 //+1901+1902+1903 g_nManualSteps[Z_AXIS] [4byte unsigned long]

#define EPR_RF_MOD_ZOS_SCAN_POINT_X       1904 //g_ZOSTestPoint[0] [1byte unsigned char]
#define EPR_RF_MOD_ZOS_SCAN_POINT_Y       1905 //g_ZOSTestPoint[1] [1byte unsigned char]

#define EPR_RF_MOD_SENSEOFFSET_OFFSET_MAX 1906 //+1907 g_nSensiblePressureOffsetMax [2byte short]
#define EPR_RF_MOD_SENSEOFFSET_DIGITS     1908 //+1909 g_nSensiblePressureDigits [2byte short]

#define EPR_RF_EMERGENCYPAUSEDIGITSMIN    1910 //+1911+1912+1913 g_nEmergencyPauseDigitsMin [4byte long]
#define EPR_RF_EMERGENCYPAUSEDIGITSMAX    1914 //+1915+1916+1917 g_nEmergencyPauseDigitsMax [4byte long]

#define EPR_RF_EMERGENCYZSTOPDIGITSMIN    1918 //+1919 g_nZEmergencyStopAllMin [2byte long]
#define EPR_RF_EMERGENCYZSTOPDIGITSMAX    1920 //+1921 g_nZEmergencyStopAllMax [2byte long]
//Nibbels: Computechecksum geht bis 2047

#define EEPROM_EXTRUDER_OFFSET          200

// bytes per extruder needed, leave some space for future development
#define EEPROM_EXTRUDER_LENGTH          100

// Extruder positions relative to extruder start
#define EPR_EXTRUDER_STEPS_PER_MM       0
#define EPR_EXTRUDER_MAX_FEEDRATE       4

// Feedrate from halted extruder in mm/s
#define EPR_EXTRUDER_MAX_START_FEEDRATE 8

// Acceleration in mm/s^2
#define EPR_EXTRUDER_MAX_ACCELERATION   12
#define EPR_EXTRUDER_HEAT_MANAGER       16
#define EPR_EXTRUDER_DRIVE_MAX          17
#define EPR_EXTRUDER_PID_PGAIN          18
#define EPR_EXTRUDER_PID_IGAIN          22
#define EPR_EXTRUDER_PID_DGAIN          26
#define EPR_EXTRUDER_PID_MAX            30
#define EPR_EXTRUDER_X_OFFSET           31
#define EPR_EXTRUDER_Y_OFFSET           35
#define EPR_EXTRUDER_WATCH_PERIOD       39
#define EPR_EXTRUDER_ADVANCE_K          41
#define EPR_EXTRUDER_DRIVE_MIN          45
#define EPR_EXTRUDER_ADVANCE_L          46
#define EPR_EXTRUDER_WAIT_RETRACT_TEMP  50
#define EPR_EXTRUDER_WAIT_RETRACT_UNITS 52
#define EPR_EXTRUDER_COOLER_SPEED       54 //byte
#define EPR_EXTRUDER_SENSOR_TYPE        94 //uint8 
#define EPR_EXTRUDER_Z_OFFSET           95 //int32_t -> +96+97+98

class EEPROM
{
#if EEPROM_MODE!=0
    static uint8_t computeChecksum();
    static void writeExtruderPrefix(uint pos);
    static void writeFloat(uint pos,PGM_P text,uint8_t digits=3);
    static void writeLong(uint pos,PGM_P text);
    static void writeInt(uint pos,PGM_P text);
    static void writeByte(uint pos,PGM_P text);
#endif // EEPROM_MODE!=0

public:
    static void init();
    static void initBaudrate();
    static void storeDataIntoEEPROM(uint8_t corrupted=0);
    static void updateChecksum();
    static void readDataFromEEPROM();
    static void restoreEEPROMSettingsFromConfiguration();
    static void initializeAllOperatingModes();
    static void clearEEPROM();
    static void writeSettings();
    static void update(GCode *com);
    static void updatePrinterUsage();
    static int getExtruderOffset(uint8_t extruder=0);

}; // EEPROM


#endif // EEPROM_H
